### PR TITLE
Added a new detector for catching unsupported COPY command structure

### PR DESCRIPTION
**Pull request by ShivanshGahlot**
### Describe the changes in this pull request
Reporting two unsupported query constructs related to COPY command:
- `COPY FROM ... WHERE`
- `COPY ... ON_ERROR`
https://github.com/yugabyte/yb-voyager/issues/1991

Added a new detector `CopyCommandUnsupportedConstructsDetector` in `detectors.go` to report both these constructs

Currently the PR also contains changes related to pg_query_go v6 upgrades. Will rebase the PR to main once I merge that PR

### Describe if there are any user-facing changes

The assessment report now reports these constructs under UQC:
image

### How was this pull request tested?

Added new unit tests for these copy constructs in:
`parser_issue_detector_test.go`
`issue_dml_test.go`
`detectors_test.go `

An end to end test for the same construct needs to be added

### Does your PR have changes that can cause upgrade issues? 
| Component        | Breaking changes? |
| :----------------------------------------------: | :-----------: |
| MetaDB                                                    |  No |
| Name registry json                                        |  No |
| Data File Descriptor Json                                 |  No |
| Export Snapshot Status Json                               |  No |
| Import Data State                                         |  No |
| Export Status Json                                        |  No |
| Data .sql files of tables                                 |  No |
| Export and import data queue                              | No |
| Schema Dump                                               | No |
| AssessmentDB                                              |  No |
| Sizing DB                                                 |  No |
| Migration Assessment Report Json                          |  No |
| Callhome Json                                             |  No |
| YugabyteD Tables                                          |  No |
| TargetDB Metadata Tables                                  |  No |

### CODE REVIEW

File: yb-voyager/src/query/queryissue/detectors.go
Comment (makalaaneesh) [inline yb-voyager/src/query/queryissue/detectors.go:210]: Simplified code from what I was discussing in the morning: 
I tried running your tests and they all passed, but do take a deeper look! 
cc: @sanyamsinghal @priyanshi-yb 
```
	protoMsg := msg.Interface().(proto.Message)
	copyStmt := protoMsg.(*pg_query.CopyStmt)

	if copyStmt.IsFrom && copyStmt.WhereClause != nil {
		d.copyFromWhereConstructDetected = true
	}
	for _, option := range copyStmt.Options {
		defElemNode := option.GetDefElem()
		if defElemNode != nil {
			if defElemNode.Defname == "on_error" {
				d.copyOnErrorConstructDetected = true
				break
			}
		}
	}

```

Comment (priyanshi-yb) [reply yb-voyager/src/query/queryissue/detectors.go:210]: I think yes this makes sense. It reduces the number of lines and readable.

Comment (sanyamsinghal) [reply yb-voyager/src/query/queryissue/detectors.go:210]: Lets stick to protomsg for this codepath, reason being the required info is not directly availabe in the struct and extraction of defElem can be generic and reused multiple times.

It can be simplified like this using existing helper functions:

```
FromField := GetMessageField(msg, "is_from")
whereField := GetMessageField(msg, "where_clause")
if FromField != nil && whereField != nil {
     d.copyFromWhereConstructDetected = true
}

defNames := TraverseAndExtractDefNamesFromDefElem(msg)
if slices.Contains(defNames, "on_error") {
    d.copyOnErrorConstructDetected = true
}

```
Refer my [PR](https://github.com/yugabyte/yb-voyager/pull/2108/files#:~:text=func%20TraverseAndExtractDefNamesFromDefElem(msg%20protoreflect.Message)%20%5B%5Dstring) for `TraverseAndExtractDefNamesFromDefElem`

**Review by sanyamsinghal — CHANGES_REQUESTED** [review]

**Comment by sanyamsinghal** [comment]
Also please add some simple end to end tests, just to cover - final assessment report is getting populated with the required info.

File: yb-voyager/src/query/queryissue/detectors.go
Comment (makalaaneesh) [reply yb-voyager/src/query/queryissue/detectors.go:218]: just to be on the same page, there are two different issues here. 

For the first one, isFrom and whereClause are directly available in `CopyStmt` struct, so would prefer dealing with the struct directly for that issue. But `GetMessageField(msg, "is_from"` isn't too bad either from a readability standpoint. I'll leave it up to you.

For the DefElem one, we can use the new function that Sanyam is introducing, but then you'll have to wait until that PR is merged.

Comment (ShivanshGahlot) [reply yb-voyager/src/query/queryissue/detectors.go:220]: I had to add a new function GetBoolField() for the is_from clause. I have used the GetMessageField and GetBoolField approach rn for the first one.
For the second one, I have used Sanyam's function in my PR.

File: yb-voyager/src/query/queryissue/parser_issue_detector_test.go
Comment (sanyamsinghal) [inline yb-voyager/src/query/queryissue/parser_issue_detector_test.go:653]: see if the tests for this can be at one place only either in detectors.go or parse_issue_detector_test.go

Currently we have the tests distributed among these two files, but since they are essentially testing the same thing, we can just have it one place.

**Review by sanyamsinghal — APPROVED** [review]
LGTM.
If you are planning to add end-to-end tests later, then do file a ticket for that.

File: yb-voyager/src/query/queryissue/parser_issue_detector_test.go
Comment (ShivanshGahlot) [reply yb-voyager/src/query/queryissue/parser_issue_detector_test.go:653]: Yeah for now I have added the tests but commented out the sql of CREATE ... ON_ERROR since it is supported only on PG17. Have added a ticket to do the same:
[https://yugabyte.atlassian.net/browse/DB-14671?atlOrigin=eyJpIjoiMTg4ZDE3NDI2MWM4NGI1MmExYzE4YzgzNjk3YjczMmMiLCJwIjoiaiJ9](url)

Comment (ShivanshGahlot) [reply yb-voyager/src/query/queryissue/parser_issue_detector_test.go:653]: Removed the tests from detectors_test.go. The tests are same anyways,

**Pull request merged**
